examples/spiffe-watcher/main.c

```diff
@@ -0,0 +1,171 @@
+/**
+ *
+ * (C) Copyright 2020-2021 Hewlett Packard Enterprise Development LP
+ *
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ *
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#include "c-spiffe/bundle/jwtbundle.h"
+#include "c-spiffe/internal/jwtutil.h"
+#include "c-spiffe/workload/workload.h"
+#include "threads.h"
+
+#include <openssl/pem.h>
+
+void callback_X509Context(workloadapi_X509Context *ctx, void *unused)
+{
+    if(ctx) {
+        x509svid_SVID **svids = ctx->svids;
+
+        for(size_t i = 0, size = arrlenu(svids); i < size; ++i) {
+            string_t str_name = spiffeid_ID_String(svids[i]->id);
+            printf("SVID updated for %s:\n", str_name);
+            arrfree(str_name);
+
+            X509 **certs = svids[i]->certs;
+            for(size_t j = 0, size = arrlenu(certs); j < size; ++j) {
+                PEM_write_X509(stdout, certs[j]);
+            }
+            putchar('\n');
+        }
+    }
+}
+
+void callback_JWTBundles(jwtbundle_Set *set, void *unused)
+{
+    if(set) {
+        jwtbundle_Bundle **bundles = jwtbundle_Set_Bundles(set);
+
+        for(size_t i = 0, size = arrlenu(bundles); i < size; ++i) {
+            printf("jwt bundle updated %s:\n", bundles[i]->td.name);
+
+            jwtutil_JWKS jwks
+                = { .root = NULL,
+                    .jwt_auths = jwtutil_CopyJWTAuthorities(bundles[i]->auths),
+                    .x509_auths = NULL };
+            err_t err;
+            string_t str = jwtutil_JWKS_Marshal(&jwks, &err);
+
+            if(str && err == NO_ERROR) {
+                printf("%s\n", str);
+            }
+
+            arrfree(str);
+            jwtutil_JWKS_Free(&jwks);
+        }
+
+        for(size_t i = 0, size = arrlenu(bundles); i < size; ++i) {
+            jwtbundle_Bundle_Free(bundles[i]);
+        }
```

Comment:
```suggestion
        jwtbundle_Set_Free(new_set);
```
Double free error, these are already freed in workload/client.cc:828.
bundles are supposed to be copied if you need them.

alternatively, you could clone the set before getting the bundles for it (see other comment), and free the new set here

Reply:
It is enough to just free the array of bundle pointers.